### PR TITLE
std::size_t modifications for module common

### PR DESCRIPTION
**Pull request by kunaltyagi**
* Replaced `size_t` with `std::size_t`
* Used modern for loops
* Used `std::size_t` for indices (in function internals) which don't affect class/struct layout
* Some changes to API to replace `int` with `std::size_t` for dimensions and indices
* Used `auto` to hide unimportant types

The aim is to to use `std::size_t` instead of `int` in all (most) modules and then convert the indices to PCL_INDEX_TYPE which can be set as `uint16_t` or `std::size_t` or even `int`. This is first in a series of such commits

### CODE REVIEW

**Comment by taketwo** [comment]
Hey, thanks for taking the initiative. The bullet-list summary is quite helpful in understanding the contents of this PR. However, it would be even greater if each of these bullet points was a separate commit. Such an approach makes the review process more manageable. For example, the first change is trivial, so it's enough to quickly scroll through such commit. Item three, however, potentially affects class/struct layout, and requires a careful review.

I consider this a good style to add outstanding TODO items to the PR descriptions. However, you should clearly state whether it's a TODO for future PRs, or you plan to update the present one. (In the latter case you can also mark it as "draft".)

**Comment by kunaltyagi** [comment]
> Item three, however, potentially affects class/struct layout, and requires a careful review.

As I mentioned, no struct level changes. I've kept the changes to inside functions

> if each of these bullet points was a separate commit

Sure. I can manage that.

> you plan to update the present one

Hopefully the present one. One PR per module, starting with 'common', then 'features' and 'filters'. Then I dunno. Stress testing is the only way I have apart from reading every line of code 😅  so some things get past my grep and segfaults

**Comment by taketwo** [comment]
> > Item three, however, potentially affects class/struct layout, and requires a careful review.
> As I mentioned, no struct level changes. I've kept the changes to inside functions

One of the purposes of the review is to verify your claim :wink:

> Hopefully the present one.

:+1:

**Comment by kunaltyagi** [comment]
The issue in filters seems to be `copyPointCloud<PointT, PointT>(*cloud_in, cloud_out)`. Can I just change it to `cloud_out = *cloud_in;` and everything would be copied, right? Does the copy function doesn't have any different functionality apart from a normal copy operation? (Though changing to `copyPointCloud(*cloud_in, cloud_out)` works too)

**Comment by taketwo** [comment]
> Does the copy function doesn't have any different functionality apart from a normal copy operation?

In the case of matching point types `copyPointCloud()` is just a `memcpy` of point data. (Plus a copy of meta-data, of course).

File: common/include/pcl/common/impl/bivariate_polynomial.hpp [hunk 1]
```diff
@@ -113,11 +115,8 @@ pcl::BivariatePolynomialT<real>::deepCopy (const pcl::BivariatePolynomialT<real>
     gradient_x = new pcl::BivariatePolynomialT<real> ();
     gradient_y = new pcl::BivariatePolynomialT<real> ();
   }
-  real* tmpParameters1 = parameters;
-  const real* tmpParameters2 = other.parameters;
-  unsigned int noOfParameters = getNoOfParameters ();
-  for (unsigned int i=0; i<noOfParameters; i++)
-    *tmpParameters1++ = *tmpParameters2++;
+
+  std::copy_n(other.parameters, getNoOfParameters (), parameters);
```
Comment (taketwo) [inline common/include/pcl/common/impl/bivariate_polynomial.hpp]: So neat now! :slightly_smiling_face:

File: common/include/pcl/common/impl/centroid.hpp [hunk 2]
```diff
@@ -88,11 +88,11 @@ pcl::compute3DCentroid (const pcl::PointCloud<PointT> &cloud,
   // If the data is dense, we don't need to check for NaN
   if (cloud.is_dense)
   {
-    for (std::size_t i = 0; i < cloud.size (); ++i)
+    for (const auto point: cloud)
```
Comment (taketwo) [inline common/include/pcl/common/impl/centroid.hpp]: `const auto& point` to avoid copy? (Also below)

File: common/include/pcl/common/impl/eigen.hpp [hunk 3]
```diff
@@ -859,8 +861,7 @@ pcl::transformPlane (const pcl::ModelCoefficients::Ptr plane_in,
   Eigen::Matrix < Scalar, 4, 1 > v_plane_in (values.data ());
   pcl::transformPlane (v_plane_in, v_plane_in, transformation);
   plane_out->values.resize (4);
-  for (int i = 0; i < 4; i++)
-    plane_in->values[i] = v_plane_in[i];
+  std::copy_n(v_plane_in.data (), 4, plane_in->values.begin ());
```
Comment (taketwo) [inline common/include/pcl/common/impl/eigen.hpp]: I know `data()` works here, but isn't `begin()` semantically more fitting?

File: common/include/pcl/common/impl/io.hpp [hunk 4]
```diff
@@ -68,10 +63,11 @@ pcl::getFieldIndex (const std::string &field_name,
   fields.clear ();
   // Get the fields list
   pcl::for_each_type<typename pcl::traits::fieldList<PointT>::type>(pcl::detail::FieldAdder<PointT>(fields));
-  for (std::size_t d = 0; d < fields.size (); ++d)
-    if (fields[d].name == field_name)
-      return (static_cast<int>(d));
-  return (-1);
+  const auto result = std::find_if(fields.begin (), fields.end (),
+      [&field_name](const auto field) { return field.name == field_name; });
```
Comment (taketwo) [inline common/include/pcl/common/impl/io.hpp]: Wouldn't `const auto&` be better here?

File: common/include/pcl/common/impl/io.hpp [hunk 5]
```diff
@@ -51,13 +51,8 @@ pcl::getFieldIndex (const pcl::PointCloud<PointT> &,
...
                     std::vector<pcl::PCLPointField> &fields)
 {
-  fields.clear ();
-  // Get the fields list
-  pcl::for_each_type<typename pcl::traits::fieldList<PointT>::type>(pcl::detail::FieldAdder<PointT>(fields));
-  for (std::size_t d = 0; d < fields.size (); ++d)
-    if (fields[d].name == field_name)
-      return (static_cast<int>(d));
-  return (-1);
+  return getFieldIndex<PointT>(std::forward<decltype(field_name)>(field_name),
```
Comment (taketwo) [inline common/include/pcl/common/impl/io.hpp]: Could you please explain this?

File: common/include/pcl/common/io.h [hunk 6]
```diff
@@ -337,21 +338,9 @@ namespace pcl
...
- 
-  /** \brief Extract the indices of a given point cloud as a new point cloud
-    * \param[in] cloud_in the input point cloud dataset
-    * \param[in] indices the vector of indices representing the points to be copied from \a cloud_in
-    * \param[out] cloud_out the resultant output point cloud dataset
-    * \note Assumes unique indices.
-    * \ingroup common
-    */
-  template <typename PointT> void 
+  template <typename PointT, typename Allocator = std::allocator<int>> void
```
Comment (taketwo) [inline common/include/pcl/common/io.h]: Do I understand right that you are "merging" versions of this function with and without explicit allocator for indices?

I was wondering if `Allocator` is specific enough for template parameter name. Maybe `IndexAllocator`, so that nobody thinks that this has anything to do with points? Also, maybe add a suffix `T` in the end to match `PointT` style? I'm just testing the waters here and I am interested to hear your opinion.

File: common/include/pcl/common/impl/bivariate_polynomial.hpp (hunk 1 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/bivariate_polynomial.hpp]: No issue of size too, int vs size_t doesn't matter 😛

File: common/include/pcl/common/impl/io.hpp (hunk 5 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/io.hpp]: The function being called is right below this. Since the first variable is being ignored, no need to duplicate code. I wanted to mark this as deprecated because it doesn't make sense but maybe later

File: common/include/pcl/common/impl/io.hpp (hunk 4 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/io.hpp]: Please do keep an eye out for such issues

File: common/include/pcl/common/impl/eigen.hpp (hunk 3 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/eigen.hpp]: `v_plane_in` is a `Matrix` and doesn't offer `begin` (I think I tried `begin` but had a compilation error)

File: common/include/pcl/common/impl/io.hpp (hunk 5 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/io.hpp]: I'm sending the parameters to the function right below this one. This function has a dead variable (which made me almost add deprecated because it feels so), so I decided to keep the live code in the function with least parameters (and same name, and function body)

File: common/include/pcl/common/io.h (hunk 6 above)
Comment (kunaltyagi) [reply common/include/pcl/common/io.h]: `IndexAllocator` sounds good. `IndexAllocatorT` implies of type T which it is not. As for being specific, I just copied it from the definition of `std::vector`

**Comment by kunaltyagi** [comment]
I accidentally rebased one commit (with `const auto&`). When you're done with the first 2 commits, please ping me, so I can push the other changes regarding renaming the template

I'm posting the diff manually for the accidentally rebased commit (apologies in advance)
```diff
diff --git a/common/include/pcl/common/impl/centroid.hpp b/common/include/pcl/common/impl/centroid.hpp
index 6bf892550..1e5a76e32 100644
--- a/common/include/pcl/common/impl/centroid.hpp
+++ b/common/include/pcl/common/impl/centroid.hpp
@@ -88,7 +88,7 @@ pcl::compute3DCentroid (const pcl::PointCloud<PointT> &cloud,
   // If the data is dense, we don't need to check for NaN
   if (cloud.is_dense)
   {
-    for (const auto point: cloud)
+    for (const auto& point: cloud)
     {
       centroid[0] += point.x;
       centroid[1] += point.y;
@@ -101,7 +101,7 @@ pcl::compute3DCentroid (const pcl::PointCloud<PointT> &cloud,
   }
   // NaN or Inf values could exist => check for them
   unsigned cp = 0;
-  for (const auto point: cloud)
+  for (const auto& point: cloud)
   {
     // Check if the point is invalid
     if (!isFinite (point))
@@ -132,7 +132,7 @@ pcl::compute3DCentroid (const pcl::PointCloud<PointT> &cloud,
   // If the data is dense, we don't need to check for NaN
   if (cloud.is_dense)
   {
-    for (const int &index : indices)
+    for (const int& index : indices)
     {
       centroid[0] += cloud[index].x;
       centroid[1] += cloud[index].y;
@@ -144,7 +144,7 @@ pcl::compute3DCentroid (const pcl::PointCloud<PointT> &cloud,
   }
   // NaN or Inf values could exist => check for them
     unsigned cp = 0;
-  for (const int &index : indices)
+  for (const int& index : indices)
   {
     // Check if the point is invalid
     if (!isFinite (cloud [index]))
diff --git a/common/include/pcl/common/impl/io.hpp b/common/include/pcl/common/impl/io.hpp
index fe170ae68..9de4b9534 100644
--- a/common/include/pcl/common/impl/io.hpp
+++ b/common/include/pcl/common/impl/io.hpp
@@ -64,7 +64,7 @@ pcl::getFieldIndex (const std::string &field_name,
   // Get the fields list
   pcl::for_each_type<typename pcl::traits::fieldList<PointT>::type>(pcl::detail::FieldAdder<PointT>(fields));
   const auto result = std::find_if(fields.begin (), fields.end (),
-      [&field_name](const auto field) { return field.name == field_name; });
+      [&field_name](const auto& field) { return field.name == field_name; });
   if (result == fields.end ())
     return -1;
   return std::distance(fields.begin (), result);
diff --git a/common/include/pcl/conversions.h b/common/include/pcl/conversions.h
index 9a4c40e3e..62f9eb6ff 100644
--- a/common/include/pcl/conversions.h
+++ b/common/include/pcl/conversions.h
@@ -314,7 +314,7 @@ namespace pcl
   inline void
   toPCLPointCloud2 (const pcl::PCLPointCloud2& cloud, pcl::PCLImage& msg)
   {
-    const auto predicate = [](const auto field) { return field.name == "rgb"; };
+    const auto predicate = [](const auto& field) { return field.name == "rgb"; };
     const auto result = std::find_if(cloud.fields.cbegin (), cloud.fields.cend (), predicate);
     if (result == cloud.fields.end ())
       throw std::runtime_error ("No rgb field!!");
diff --git a/common/src/PCLPointCloud2.cpp b/common/src/PCLPointCloud2.cpp
index d3537a15d..85bdaa1af 100644
--- a/common/src/PCLPointCloud2.cpp
+++ b/common/src/PCLPointCloud2.cpp
@@ -152,7 +152,7 @@ pcl::PCLPointCloud2::concatenate (pcl::PCLPointCloud2 &cloud1, const pcl::PCLPoi
   }
   for (std::size_t cp = 0; cp < size2; ++cp)
   {
-    for (const auto field_data: valid_fields)
+    for (const auto& field_data: valid_fields)
     {
       const auto& i = field_data.idx1;
       const auto& j = field_data.idx2;

```

File: common/include/pcl/common/impl/io.hpp (hunk 5 above)
Comment (taketwo) [reply common/include/pcl/common/impl/io.hpp]: OK, got that. But what is `std::forward<decltype(...)>` for? Wouldn't it work without?

File: common/include/pcl/common/io.h (hunk 6 above)
Comment (taketwo) [reply common/include/pcl/common/io.h]: > As for being specific, I just copied it from the definition of std::vector

Yeah, this is understood. But this function has more templated entities, so further clarification is needed IMHO.

> XAllocatorT implies type T which it is not (for the allocator).

Hm, I always assumed that `T` is just short for "type". That is, `PointT` means "point type". But anaway, I'm totally fine without `T`. This also matches STL style.

> IndicesVecAllocator sounds good to me.

Or `IndicesVectorAllocator`. But up to you.

**Comment by taketwo** [comment]
No worries. Yes, I think I'm done with the first two.

File: common/include/pcl/common/io.h (hunk 6 above)
Comment (kunaltyagi) [reply common/include/pcl/common/io.h]: > PointT means "point type"

I always thought `PointT` was shorthand for type T but it's used for points (`PointRGBA`, `PointXYZ`).

Comment (taketwo) [reply common/include/pcl/common/io.h]: Well, perhaps you are right. But we'll never know for sure :smile:

File: common/include/pcl/common/impl/io.hpp (hunk 5 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/io.hpp]: `std::forward` allows a temporary to remain one and not become a reference. Since the `field_name` might be just constant char array (like "x"), the compiler might create std::string and pass its reference instead of passing it directly.

To be frank, it might not affect anything here, but it's a reflex to forward arguments using `std::forward` to prevent a copy (and for correct selection of overloads, lest someone take reference to a temporary because we passed a reference variable, not a temporary)

Comment (taketwo) [reply common/include/pcl/common/impl/io.hpp]: > Since the field_name might be just constant char array (like "x"),

I don't think I agree with you here. `field_name` **is** a `const std::string&`, no matter how the function was called. Yes, it may be bound to a temporary constructed out of a C-string, but still the type of `field_name` variable is fixed to be `const std::string&`. Therefore, `decltype(field_name)` === `const std::string&`.

File: common/src/pcl_base.cpp [hunk 7]
```diff
@@ -70,7 +70,7 @@ pcl::PCLBase<pcl::PCLPointCloud2>::setInputCloud (const PCLPointCloud2ConstPtr &
 
   // Obtain the size of all fields. Restrict to sizeof FLOAT32 for now
   field_sizes_.resize (input_->fields.size ());
-  for (std::size_t d = 0; d < input_->fields.size (); ++d)
+  for (int d = 0; d < static_cast<int>(field_sizes_.size ()); ++d)
```
Comment (taketwo) [inline common/src/pcl_base.cpp]: Can you explain this change?

Comment (kunaltyagi) [reply common/src/pcl_base.cpp]: The assumption is that `fields.size() < INT_MAX`, so there's no need to make that loop using `std::size_t`. We can codify it. `using PointSizeT = std::size_t; using FieldSizeT = int`

**Comment by kunaltyagi** [comment]
Oops, pressed the wrong button. Vim plugins for browsers should have an undo 😢

File: common/src/pcl_base.cpp (hunk 7 above)
Comment (taketwo) [reply common/src/pcl_base.cpp]: Well true, I think it's common sense to assume we have at most couple dozen fields in a point cloud. But is `fields.size() < INT_MAX` explicitly stated anywhere? And even if so, I'm not sure explicit cast to `int` makes sense. It does not _enforce_ the convention, rather breaks silently if the number of fields is too large.

Comment (kunaltyagi) [reply common/src/pcl_base.cpp]: AFAIK, no. But In several other places, this silent breaking on contract is done. Maybe enforce it in PCLField or related class?

**Comment by SergioRAgostinho** [comment]
>   * Replaced `size_t` with `std::size_t`

What's the rationale on this on this?

**Comment by kunaltyagi** [comment]
> What's the rationale on this?

My understanding of the standard says that the header `<cstddef>` may or may not declare `::size_t`, so it cannot be relied upon for making sure that `::size_t` is present or absent, unless specifically including `<stddef.h>` or another header from the C library which is guaranteed to declare it.

Since we are slowly migrating away from C functions, there's no guarantee that the all stdlib will provide the optional declaration of `::size_t`. As such it doesn't make sense to keep using `size_t` when we actually mean `std::size_t`.

We can declare in `pcl` namespace `using std::size_t` as an alternative

**Comment by taketwo** [comment]
TL;DR: same reason as migrating from `abs` to `std::abs`. Former is a legacy C code, and the latter is proper C++.

File: common/src/pcl_base.cpp (hunk 7 above)
Comment (taketwo) [reply common/src/pcl_base.cpp]: > Maybe enforce it in PCLField or related class?

Perhaps, but please elaborate on what you have in mind.

Comment (kunaltyagi) [reply common/src/pcl_base.cpp]: stdlib enforces that `size()` returns `std::size_t`. So, we'll need to `static_cast`. One way to enforce the size limit would be the places where I've seen fields being added:
`POINT_CLOUD_REGISTER_POINT_STRUCT`, `FieldAdder` and `concatenateFields`.

The first one is pure-preprocessor but I don't see a way to limit the total items in a preprocessor friendly manner.

But we can enforce the `INT_MAX` or `UINT_MAX` in `conversions.h`, `io.hpp` and `io.cpp` for the other two structs/functions. Enforcing would just involve throwing an exception because that's a really extraordinary situation

**Comment by kunaltyagi** [comment]
I'll check this code for use of reserve before proceeding

File: common/src/pcl_base.cpp (hunk 7 above)
Comment (kunaltyagi) [reply common/src/pcl_base.cpp]: Nothing except the fact that it's not needed (now 😝 ). Maybe using uint64 makes sense because I don't think the headaches of 2 separate max_range makes sense for the measly space savings the small number of field indices will give (We'll use 50% more memory 🤷‍♂ of a small vector)

**Comment by kunaltyagi** [comment]
No changes, just a rebase to bring everything up to date (MacOS CI, conflict due to `vector.reserve()`)

File: common/include/pcl/common/impl/io.hpp [hunk 8]
```diff
@@ -51,13 +51,7 @@ pcl::getFieldIndex (const pcl::PointCloud<PointT> &,
                     const std::string &field_name,
```
Comment (SergioRAgostinho) [inline common/include/pcl/common/impl/io.hpp]: This function is a good candidate for deprecation.

File: common/src/pcl_base.cpp (hunk 7 above)
Comment (kunaltyagi) [reply common/src/pcl_base.cpp]: So, should I go ahead and remove cast and assumption of int from `field_indices.size()`?

File: common/include/pcl/common/pca.h [hunk 9]
```diff
@@ -166,7 +166,7 @@ namespace pcl
         * \param[in] nb_cols the number of columns to be considered col_start included
         */
       void
-      setIndices (size_t row_start, size_t col_start, size_t nb_rows, size_t nb_cols) override
+      setIndices (std::size_t row_start, size_t col_start, size_t nb_rows, size_t nb_cols) override
```
[the remark this answers is not in the document]
Comment (taketwo) [reply common/include/pcl/common/pca.h]: @kunaltyagi If you were using `s/size_t/std::size_t/` then you have probably forgotten to add `g` in the end. If so, please double-check with other files as well.

File: common/src/pcl_base.cpp (hunk 7 above)
Comment (taketwo) [reply common/src/pcl_base.cpp]: Having had that discussion, I'd say yes.

File: common/include/pcl/common/impl/io.hpp (hunk 8 above)
Comment (kunaltyagi) [reply common/include/pcl/common/impl/io.hpp]: I can see a usecase when `auto` is used, but usually, I've seen `PointT` being a separate template parameter. So it can definitely be deprecated (especially since `decltype(cloud)::PointType` is available)

File: common/include/pcl/common/pca.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (kunaltyagi) [reply common/include/pcl/common/pca.h]: In kiss_fft.h, I found:
```
#ifdef __cplusplus
extern "C" {
#endif
```
What should I perform in this file? Let `size_t` remain? Assume C compiler will not be used, and remove it in a separate commit?

[the remark this answers is not in the document]
Comment (taketwo) [reply common/include/pcl/common/pca.h]: I would simply remove the file :laughing:

But seriously, don't know. Maybe let `size_t` remain?

**Comment by SergioRAgostinho** [comment]
I accepted the changes, however now one of the tutorials in failing to build. I'm unsure if is related to this PR though.

**Comment by taketwo** [comment]
Yep, `vhf_recognition` tutorial is using the now-deprecated function, and warnings are treated as errors.

I think it should be a rule that when we deprecate something, we grep through the codebase and remove usage of deprecated functions from PCL itself. (This, by the way, has not been done for the previous deprecation, see #3349.)

**Comment by kunaltyagi** [comment]
PTAL

**Comment by taketwo** [comment]
Thanks. The new separation into commits makes sense to me, except for one case. In the "Letting the compiler auto-deduce template parameters" commit the changes to `io.h` and `io.hpp` do not belong there I think. These changes are updating the lines touched by "Better loops via modern for, algorithms or redirection" and thus should be squashed with it.

**Comment by kunaltyagi** [comment]
Another flaky test found:

```
[----------] 1 test from SacTest/1, where TypeParam = pcl::LeastMedianSquares<pcl::PointXYZ>
111: [ RUN      ] SacTest/1.InfiniteLoop
111: /__w/1/s/test/sample_consensus/test_sample_consensus.cpp:134: Failure
111: Value of: cv.wait_for (lock, 1s)
111:   Actual: 4-byte object <01-00 00-00>
111: Expected: std::cv_status::no_timeout
111: Which is: 4-byte object <00-00 00-00>
111: [  FAILED  ] SacTest/1.InfiniteLoop, where TypeParam = pcl::LeastMedianSquares<pcl::PointXYZ> (1001 ms)
111: [----------] 1 test from SacTest/1 (1001 ms total)
```

**Comment by kunaltyagi** [comment]
Fixed the separation issue after a mistake 😭 . PTAL
